docs/index.rst

```diff
@@ -3,8 +3,128 @@
    You can adapt this file completely to your liking, but it should at least
    contain the root `toctree` directive.
 
-Welcome to uarray's documentation!
-==================================
+Welcome to ``uarray``'s documentation
+=====================================
+
+.. warning::
+    ``uarray`` is a developer tool, it is not meant to be used directly by end-users.
+
+.. warning::
+    This document is meant to elicit discussion from the broader community and to help
+    drive the direction that ``uarray`` goes towards. Examples proviced here may not be
+    immediately stable.
+
+``uarray`` is a backend/dispatch mechanism with a focus on array computing and the
+needs of the wider array community. It is meant to address the shortcomings of `NEP-18
+<http://www.numpy.org/neps/nep-0018-array-function-protocol.html>`_ and `NEP-13
+<https://www.numpy.org/neps/nep-0013-ufunc-overrides.html>`_; while also being a
+comprehensive solution to `NEP-22 <http://www.numpy.org/neps/nep-0022-ndarray-duck-typing-overview.html>`_,
+while still holding nothing in ``uarray`` itself that's specific to array computing
+or the NumPy API.
+
+Design Philosophies
+-------------------
+
+The following section discusses the design philosophies of ``uarray``, and the
+reasoning behind some of these philosophies.
+
+Modularity
+^^^^^^^^^^
+
+``uarray`` (and its sister modules ``unumpy`` and others to come) were designed
+from the ground-up to be modular. This is part of why ``uarray`` itself holds
+the core backend and dispatch machinery, and ``unumpy`` holds the actual
+multimethods. Also, ``unumpy`` can be developed completely separately to
+``uarray``, although the ideal place to have it would be NumPy itself.
+
+However, the benefit to having it separate is that it could span multiple
+NumPy versions, even before NEP-18 (or even NEP-13) was available.
+
+Separate Imports
+^^^^^^^^^^^^^^^^
+
+Code wishing to use the backend machinery for NumPy (as an example) will
+use the statement ``import unumpy as np`` instead of the usual
+``import numpy as np``. This is deliberate: it makes dispatching opt-in
+instead of being forced to use it, and the overhead associated with it.
+However, a package is free to define its main methods as the dispatchable
+versions, thereby allowing dispatch on the default implementation.
+
+Extensibility *and* Choice
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+If some effort is put into the dispatch machinery, it's possible to
+dispatch over arbitrary objects --- including arrays, dtypes, and
+so on. A method defines the type of each dispatchable argument, and
+backends are *only* passed types they know how to dispatch over.
```

Comment:
This brings up some questions for me:

- How do you deal with functions that allow flexible inputs like `array_like`? 
- Does this lead to the need for using templating to cover a decent range of types? And does it suffer from combinatorial explosion, like when using Tempita + Cython?

Reply:
So, my view is that `array_like` tries to do too much. What I've done is register just `ndarray` and allow coercion if someone really, really wants to use `array_like`.

Reply:
I've also registered common things like `tuple, list, int, float, bool`, et. al. as "fallbacks", which means if they're encountered, NumPy will be tried again, but last.

Reply:
Ah, this part: It means whether it'll be asked to decide based on arrays, dtypes or ufuncs. I've made that more clear.